File: SDK/nncamsdk/dotnet/samples/demowinformcs2/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Imaging;
using System.Text;
using System.Windows.Forms;
using System.Runtime.InteropServices;

namespace demowinformcs2
{
    public partial class Form1 : Form
    {
        private delegate void DelegateEvent(Toupcam.eEVENT[] ev);
        private Toupcam cam_ = null;
        private Bitmap bmp_ = null;
        private DelegateEvent ev_ = null;

        private void OnEventError()
        {
            if (cam_ != null)
            {
                cam_.Close();
                cam_ = null;
            }
            MessageBox.Show("Error");
        }

        private void OnEventDisconnected()
        {
            if (cam_ != null)
            {
                cam_.Close();
                cam_ = null;
            }
            MessageBox.Show("The camera is disconnected, maybe has been pulled out.");
        }

        private void OnEventExposure()
        {
            if (cam_ != null)
            {
                uint nTime = 0;
                if (cam_.get_ExpoTime(out nTime))
                {
                    trackBar1.Value = (int)nTime;
                    label1.Text = (nTime / 1000).ToString() + " ms";
                }
            }
        }

        private void OnEventImage()
        {
            if (bmp_ != null)
            {
                BitmapData bmpdata = bmp_.LockBits(new Rectangle(0, 0, bmp_.Width, bmp_.Height), ImageLockMode.WriteOnly, bmp_.PixelFormat);

                Toupcam.FrameInfoV2 info = new Toupcam.FrameInfoV2();
                cam_.PullImageV2(bmpdata.Scan0, 24, out info);

                bmp_.UnlockBits(bmpdata);

                pictureBox1.Image = bmp_;
                pictureBox1.Invalidate();
            }
        }

        private void OnEventStillImage()
        {
            Toupcam.FrameInfoV2 info = new Toupcam.FrameInfoV2(); ;
            if (cam_.PullStillImageV2(IntPtr.Zero, 24, out info))   /* peek the width and height */
            {
                Bitmap sbmp = new Bitmap((int)info.width, (int)info.height, PixelFormat.Format24bppRgb);

                BitmapData bmpdata = sbmp.LockBits(new Rectangle(0, 0, sbmp.Width, sbmp.Height), ImageLockMode.WriteOnly, sbmp.PixelFormat);
                cam_.PullStillImageV2(bmpdata.Scan0, 24, out info);
                sbmp.UnlockBits(bmpdata);

                sbmp.Save("demowinformcs2.jpg");
            }
        }

        public Form1()
        {
            InitializeComponent();
            pictureBox1.Width = ClientRectangle.Right - button1.Bounds.Right - 20;
            pictureBox1.Height = ClientRectangle.Height - 8;
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            button2.Enabled = false;
            button3.Enabled = false;
            trackBar1.Enabled = false;
            trackBar2.Enabled = false;
            trackBar3.Enabled = false;
            checkBox1.Enabled = false;
            comboBox1.Enabled = false;
        }

        private void DelegateOnEvent(Toupcam.eEVENT[] ev)
        {
            switch (ev[0])
            {
                case Toupcam.eEVENT.EVENT_ERROR:
                    OnEventError();
                    break;
                case Toupcam.eEVENT.EVENT_DISCONNECTED:
                    OnEventDisconnected();
                    break;
                case Toupcam.eEVENT.EVENT_EXPOSURE:
                    OnEventExposure();
                    break;
                case Toupcam.eEVENT.EVENT_IMAGE:
                    OnEventImage();
                    break;
                case Toupcam.eEVENT.EVENT_STILLIMAGE:
                    OnEventStillImage();
                    break;
                case Toupcam.eEVENT.EVENT_TEMPTINT:
                    OnEventTempTint();
                    break;
            }
        }

        private void DelegateOnEventCallback(Toupcam.eEVENT ev)
        {
            /* this delegate is call by internal thread of toupcam.dll which is NOT the same of UI thread.
             * Why we use BeginInvoke, Please see:
             * http://msdn.microsoft.com/en-us/magazine/cc300429.aspx
             * http://msdn.microsoft.com/en-us/magazine/cc188732.aspx
             * http://stackoverflow.com/questions/1364116/avoiding-the-woes-of-invoke-begininvoke-in-cross-thread-winform-event-handling
            */
            BeginInvoke(ev_, new Toupcam.eEVENT[1] { ev });
        }

        private void OnStart(object sender, EventArgs e)
        {
            if (cam_ != null)
                return;

            Toupcam.DeviceV2[] arr = Toupcam.EnumV2();
            if (arr.Length <= 0)
                MessageBox.Show("no device");
            else
            {
                cam_ = Toupcam.Open(arr[0].id);
                if (cam_ != null)
                {
                    checkBox1.Enabled = true;
                    trackBar1.Enabled = true;
                    trackBar2.Enabled = true;
                    trackBar3.Enabled = true;
                    comboBox1.Enabled = true;
                    button2.Enabled = true;
                    button3.Enabled = true;
                    button2.ContextMenuStrip = null;
                    InitSnapContextMenuAndExpoTimeRange();

                    trackBar2.SetRange(2000, 15000);
                    trackBar3.SetRange(200, 2500);
                    OnEventTempTint();

                    uint resnum = cam_.ResolutionNumber;
                    uint eSize = 0;
                    if (cam_.get_eSize(out eSize))
                    {
                        for (uint i = 0; i < resnum; ++i)
                        {
                            int w = 0, h = 0;
                            if (cam_.get_Resolution(i, out w, out h))
                                comboBox1.Items.Add(w.ToString() + "*" + h.ToString());
                        }
                        comboBox1.SelectedIndex = (int)eSize;

                        int width = 0, height = 0;
                        if (cam_.get_Size(out width, out height))
                        {
                            bmp_ = new Bitmap(width, height, PixelFormat.Format24bppRgb);
                            ev_ = new DelegateEvent(DelegateOnEvent);
                            if (!cam_.StartPullModeWithCallback(new Toupcam.DelegateEventCallback(DelegateOnEventCallback)))
                                MessageBox.Show("failed to start device");
                            else
                            {
                                bool autoexpo = true;
                                cam_.get_AutoExpoEnable(out autoexpo);
                                checkBox1.Checked = autoexpo;
                                trackBar1.Enabled = !checkBox1.Checked;
                            }
                        }
                    }
                }
            }
        }

        private void SnapClickedHandler(object sender, ToolStripItemClickedEventArgs e)
        {
            int k = button2.ContextMenuStrip.Items.IndexOf(e.ClickedItem);
            if (k >= 0)
                cam_.Snap((uint)k);
        }

        private void InitSnapContextMenuAndExpoTimeRange()
        {
            if (cam_ == null)
                return;

            uint nMin = 0, nMax = 0, nDef = 0;
            if (cam_.get_ExpTimeRange(out nMin, out nMax, out nDef))
                trackBar1.SetRange((int)nMin, (int)nMax);
            OnEventExposure();

            if (cam_.StillResolutionNumber <= 0)
                return;
            
            button2.ContextMenuStrip = new ContextMenuStrip();
            button2.ContextMenuStrip.ItemClicked += new ToolStripItemClickedEventHandler(this.SnapClickedHandler);

            if (cam_.StillResolutionNumber < cam_.ResolutionNumber)
            {
                uint eSize = 0;
                if (cam_.get_eSize(out eSize))
                {
                    if (0 == eSize)
                    {
                        StringBuilder sb = new StringBuilder();
                        int w = 0, h = 0;
                        cam_.get_Resolution(eSize, out w, out h);
                        sb.Append(w);
                        sb.Append(" * ");
                        sb.Append(h);
                        button2.ContextMenuStrip.Items.Add(sb.ToString());
                        return;
                    }
                }
            }

            for (uint i = 0; i < cam_.ResolutionNumber; ++i)
            {
                StringBuilder sb = new StringBuilder();
                int w = 0, h = 0;
                cam_.get_Resolution(i, out w, out h);
                sb.Append(w);
                sb.Append(" * ");
                sb.Append(h);
                button2.ContextMenuStrip.Items.Add(sb.ToString());
            }
        }

        private void OnSnap(object sender, EventArgs e)
        {
            if (cam_ != null)
            {
                if (cam_.StillResolutionNumber <= 0)
                {
                    if (bmp_ != null)
                    {
                        bmp_.Save("demowinformcs2.jpg");
                    }
                }
                else
                {
                    if (button2.ContextMenuStrip != null)
                        button2.ContextMenuStrip.Show(Cursor.Position);
                }
            }
        }

        private void OnClosing(object sender, FormClosingEventArgs e)
        {
            if (cam_ != null)
            {
                cam_.Close();
                cam_ = null;
            }
        }

        private void OnSelectResolution(object sender, EventArgs e)
        {
            if (cam_ != null)
            {
                uint eSize = 0;
                if (cam_.get_eSize(out eSize))
                {
                    if (eSize != comboBox1.SelectedIndex)
                    {
                        button2.ContextMenuStrip = null;

                        cam_.Stop();
                        cam_.put_eSize((uint)comboBox1.SelectedIndex);

                        InitSnapContextMenuAndExpoTimeRange();
                        OnEventTempTint();

                        int width = 0, height = 0;
                        if (cam_.get_Size(out width, out height))
                        {
                            bmp_ = new Bitmap(width, height, PixelFormat.Format24bppRgb);
                            ev_ = new DelegateEvent(DelegateOnEvent);
                            cam_.StartPullModeWithCallback(new Toupcam.DelegateEventCallback(DelegateOnEventCallback));
                        }
                    }
                }
            }
        }

        private void checkBox1_CheckedChanged(object sender, EventArgs e)
        {
            if (cam_ != null)
                cam_.put_AutoExpoEnable(checkBox1.Checked);
            trackBar1.Enabled = !checkBox1.Checked;
        }

        private void OnExpoValueChange(object sender, EventArgs e)
        {
            if (!checkBox1.Checked)
            {
                if (cam_ != null)
                {
                    uint n = (uint)trackBar1.Value;
                    cam_.put_ExpoTime(n);
                    label1.Text = (n / 1000).ToString() + " ms";
                }
            }
        }

        private void Form_SizeChanged(object sender, EventArgs e)
        {
            pictureBox1.Width = ClientRectangle.Right - button1.Bounds.Right - 20;
            pictureBox1.Height = ClientRectangle.Height - 8;
        }

        private void OnEventTempTint()
        {
            if (cam_ != null)
            {
                int nTemp = 0, nTint = 0;
                if (cam_.get_TempTint(out nTemp, out nTint))
                {
                    label2.Text = nTemp.ToString();
                    label3.Text = nTint.ToString();
                    trackBar2.Value = nTemp;
                    trackBar3.Value = nTint;
                }
            }
        }

        private void OnWhiteBalanceOnePush(object sender, EventArgs e)
        {
            if (cam_ != null)
                cam_.AwbOnePush();
        }

        private void OnTempTintChanged(object sender, EventArgs e)
        {
            if (cam_ != null)
                cam_.put_TempTint(trackBar2.Value, trackBar3.Value);
            label2.Text = trackBar2.Value.ToString();
            label3.Text = trackBar3.Value.ToString();
        }
    }
}

File: SDK/nncamsdk/dotnet/samples/demowinformvb/Form1.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.Text
Imports System.Windows.Forms
Imports System.Runtime.InteropServices

Public Class Form1
    Private cam_ As Toupcam = Nothing
    Private bmp_ As Bitmap = Nothing
    Private MSG_CAMEVENT As UInteger = &H8001 ' WM_APP = 0x8000

    Private Sub OnEventError()
        If cam_ IsNot Nothing Then
            cam_.Close()
            cam_ = Nothing
        End If
        MessageBox.Show("Error")
    End Sub

    Private Sub OnEventDisconnected()
        If cam_ IsNot Nothing Then
            cam_.Close()
            cam_ = Nothing
        End If
        MessageBox.Show("The camera is disconnected, maybe has been pulled out.")
    End Sub

    Private Sub OnEventExposure()
        If cam_ IsNot Nothing Then
            Dim nTime As UInteger = 0
            If cam_.get_ExpoTime(nTime) Then
                trackBar1.Value = CInt(nTime)
                label1.Text = (nTime \ 1000).ToString() & " ms"
            End If
        End If
    End Sub

    Private Sub OnEventImage()
        If bmp_ IsNot Nothing Then
            Dim bmpdata As BitmapData = bmp_.LockBits(New Rectangle(0, 0, bmp_.Width, bmp_.Height), ImageLockMode.[WriteOnly], bmp_.PixelFormat)

            Dim info As New Toupcam.FrameInfoV2
            cam_.PullImageV2(bmpdata.Scan0, 24, info)

            bmp_.UnlockBits(bmpdata)

            pictureBox1.Image = bmp_
            pictureBox1.Invalidate()
        End If
    End Sub

    Private Sub OnEventStillImage()
        Dim info As New Toupcam.FrameInfoV2
        If cam_.PullStillImageV2(IntPtr.Zero, 24, info) Then ' peek the width and height 
            Dim sbmp As New Bitmap(CInt(info.width), CInt(info.height), PixelFormat.Format24bppRgb)

            Dim bmpdata As BitmapData = sbmp.LockBits(New Rectangle(0, 0, sbmp.Width, sbmp.Height), ImageLockMode.[WriteOnly], sbmp.PixelFormat)
            cam_.PullStillImageV2(bmpdata.Scan0, 24, info)
            sbmp.UnlockBits(bmpdata)

            sbmp.Save("demowinformvb.jpg")
        End If
    End Sub

    Public Sub New()
        InitializeComponent()
        pictureBox1.Width = ClientRectangle.Right - button1.Bounds.Right - 20
        pictureBox1.Height = ClientRectangle.Height - 8
    End Sub

    <System.Security.Permissions.PermissionSet(System.Security.Permissions.SecurityAction.Demand, Name:="FullTrust")> _
    Protected Overrides Sub WndProc(ByRef m As Message)
        If MSG_CAMEVENT = m.Msg Then
            Select Case m.WParam.ToInt32()
                Case Toupcam.eEVENT.EVENT_ERROR
                    OnEventError()
                    Exit Select
                Case Toupcam.eEVENT.EVENT_DISCONNECTED
                    OnEventDisconnected()
                    Exit Select
                Case Toupcam.eEVENT.EVENT_EXPOSURE
                    OnEventExposure()
                    Exit Select
                Case Toupcam.eEVENT.EVENT_IMAGE
                    OnEventImage()
                    Exit Select
                Case Toupcam.eEVENT.EVENT_STILLIMAGE
                    OnEventStillImage()
                    Exit Select
                Case Toupcam.eEVENT.EVENT_TEMPTINT
                    OnEventTempTint()
                    Exit Select
            End Select
            Return
        End If
        MyBase.WndProc(m)
    End Sub

    Private Sub SnapClickedHandler(sender As Object, e As ToolStripItemClickedEventArgs)
        Dim k As Integer = button2.ContextMenuStrip.Items.IndexOf(e.ClickedItem)
        If k >= 0 Then
            cam_.Snap(CUInt(k))
        End If
    End Sub

    Private Sub InitSnapContextMenuAndExpoTimeRange()
        If cam_ Is Nothing Then
            Return
        End If

        Dim nMin As UInteger = 0, nMax As UInteger = 0, nDef As UInteger = 0
        If cam_.get_ExpTimeRange(nMin, nMax, nDef) Then
            trackBar1.SetRange(CInt(nMin), CInt(nMax))
        End If
        OnEventExposure()

        If cam_.StillResolutionNumber <= 0 Then
            Return
        End If

        button2.ContextMenuStrip = New ContextMenuStrip()
        AddHandler button2.ContextMenuStrip.ItemClicked, AddressOf Me.SnapClickedHandler

        If cam_.StillResolutionNumber < cam_.ResolutionNumber Then
            Dim eSize As UInteger = 0
            If cam_.get_eSize(eSize) Then
                If 0 = eSize Then
                    Dim sb As New StringBuilder()
                    Dim w As Integer = 0, h As Integer = 0
                    cam_.get_Resolution(eSize, w, h)
                    sb.Append(w)
                    sb.Append(" * ")
                    sb.Append(h)
                    button2.ContextMenuStrip.Items.Add(sb.ToString())
                    Return
                End If
            End If
        End If

        For i As UInteger = 0 To cam_.ResolutionNumber - 1
            Dim sb As New StringBuilder()
            Dim w As Integer = 0, h As Integer = 0
            cam_.get_Resolution(i, w, h)
            sb.Append(w)
            sb.Append(" * ")
            sb.Append(h)
            button2.ContextMenuStrip.Items.Add(sb.ToString())
        Next
    End Sub

    Private Overloads Sub OnClosing(sender As Object, e As FormClosingEventArgs) Handles MyBase.FormClosing
        If cam_ IsNot Nothing Then
            cam_.Close()
            cam_ = Nothing
        End If
    End Sub

    Private Sub Form_SizeChanged(sender As Object, e As EventArgs) Handles MyBase.SizeChanged
        pictureBox1.Width = ClientRectangle.Right - button1.Bounds.Right - 20
        pictureBox1.Height = ClientRectangle.Height - 8
    End Sub

    Private Sub OnEventTempTint()
        If cam_ IsNot Nothing Then
            Dim nTemp As Integer = 0, nTint As Integer = 0
            If cam_.get_TempTint(nTemp, nTint) Then
                label2.Text = nTemp.ToString()
                label3.Text = nTint.ToString()
                trackBar2.Value = nTemp
                trackBar3.Value = nTint
            End If
        End If
    End Sub

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        Me.button2.Enabled = False
        Me.button3.Enabled = False
        Me.trackBar1.Enabled = False
        Me.trackBar2.Enabled = False
        Me.trackBar3.Enabled = False
        Me.checkBox1.Enabled = False
        Me.comboBox1.Enabled = False
    End Sub

    Private Sub OnStart(sender As Object, e As EventArgs) Handles button1.Click
        If cam_ IsNot Nothing Then
            Return
        End If

        Dim arr As Toupcam.DeviceV2() = Toupcam.[EnumV2]()
        If arr.Length <= 0 Then
            MessageBox.Show("no device")
        Else
            cam_ = Toupcam.Open(arr(0).id)
            If cam_ IsNot Nothing Then
                checkBox1.Enabled = True
                trackBar1.Enabled = True
                trackBar2.Enabled = True
                trackBar3.Enabled = True
                comboBox1.Enabled = True
                button2.Enabled = True
                button3.Enabled = True
                button2.ContextMenuStrip = Nothing
                InitSnapContextMenuAndExpoTimeRange()

                trackBar2.SetRange(2000, 15000)
                trackBar3.SetRange(200, 2500)
                OnEventTempTint()

                Dim resnum As UInteger = cam_.ResolutionNumber
                Dim eSize As UInteger = 0
                If cam_.get_eSize(eSize) Then
                    For i As UInteger = 0 To resnum - 1
                        Dim w As Integer = 0, h As Integer = 0
                        If cam_.get_Resolution(i, w, h) Then
                            comboBox1.Items.Add(w.ToString() & "*" & h.ToString())
                        End If
                    Next
                    comboBox1.SelectedIndex = CInt(eSize)

                    Dim width As Integer = 0, height As Integer = 0
                    If cam_.get_Size(width, height) Then
                        bmp_ = New Bitmap(width, height, PixelFormat.Format24bppRgb)
                        If Not cam_.StartPullModeWithWndMsg(Me.Handle, MSG_CAMEVENT) Then
                            MessageBox.Show("failed to start device")
                        Else
                            Dim autoexpo As Boolean = True
                            cam_.get_AutoExpoEnable(autoexpo)
                            checkBox1.Checked = autoexpo
                            trackBar1.Enabled = Not checkBox1.Checked
                        End If
                    End If
                End If
            End If
        End If
    End Sub

    Private Sub OnSnap(sender As Object, e As EventArgs) Handles button2.Click
        If cam_ IsNot Nothing Then
            If cam_.StillResolutionNumber <= 0 Then
                If bmp_ IsNot Nothing Then
                    bmp_.Save("demowinformvb.jpg")
                End If
            Else
                If button2.ContextMenuStrip IsNot Nothing Then
                    button2.ContextMenuStrip.Show(Cursor.Position)
                End If
            End If
        End If
    End Sub

    Private Sub OnSelectResolution(sender As Object, e As EventArgs) Handles comboBox1.SelectedIndexChanged
        If cam_ IsNot Nothing Then
            Dim eSize As UInteger = 0
            If cam_.get_eSize(eSize) Then
                If eSize <> comboBox1.SelectedIndex Then
                    button2.ContextMenuStrip = Nothing

                    cam_.[Stop]()
                    cam_.put_eSize(CUInt(comboBox1.SelectedIndex))

                    InitSnapContextMenuAndExpoTimeRange()
                    OnEventTempTint()

                    Dim width As Integer = 0, height As Integer = 0
                    If cam_.get_Size(width, height) Then
                        bmp_ = New Bitmap(width, height, PixelFormat.Format24bppRgb)
                        cam_.StartPullModeWithWndMsg(Me.Handle, MSG_CAMEVENT)
                    End If
                End If
            End If
        End If
    End Sub

    Private Sub checkBox1_CheckedChanged(sender As Object, e As EventArgs) Handles checkBox1.CheckedChanged
        If cam_ IsNot Nothing Then
            cam_.put_AutoExpoEnable(checkBox1.Checked)
        End If
        trackBar1.Enabled = Not checkBox1.Checked
    End Sub

    Private Sub OnExpoValueChange(sender As Object, e As EventArgs) Handles trackBar1.ValueChanged
        If Not checkBox1.Checked Then
            If cam_ IsNot Nothing Then
                Dim n As UInteger = CUInt(trackBar1.Value)
                cam_.put_ExpoTime(n)
                label1.Text = (n \ 1000).ToString() & " ms"
            End If
        End If
    End Sub

    Private Sub OnWhiteBalanceOnePush(sender As Object, e As EventArgs) Handles button3.Click
        If cam_ IsNot Nothing Then
            cam_.AwbOnePush()
        End If
    End Sub

    Private Sub OnTempTintChanged(sender As Object, e As EventArgs) Handles trackBar2.ValueChanged, trackBar3.ValueChanged
        If cam_ IsNot Nothing Then
            cam_.put_TempTint(trackBar2.Value, trackBar3.Value)
        End If
        label2.Text = trackBar2.Value.ToString()
        label3.Text = trackBar3.Value.ToString()
    End Sub

End Class